numpy/lib/arraysetops.py

```diff
@@ -173,6 +187,50 @@ def unique(ar, return_index=False, return_inverse=False, return_counts=False):
     array([1, 2, 6, 4, 2, 3, 2])
 
     """
+    ar = np.asanyarray(ar)
+    if axis is None:
+        return _unique1d(ar, return_index, return_inverse, return_counts)
+    if abs(axis) > ar.ndim:
+        raise ValueError('Invalid axis kwarg specified for unique')
+
+    ar = np.swapaxes(ar, axis, 0)
+    orig_shape, orig_dtype = ar.shape, ar.dtype
+    # Must reshape to a contiguous 2D array for this to work...
+    ar = ar.reshape(orig_shape[0], -1)
+    ar = np.ascontiguousarray(ar)
+
+    if ar.dtype.char in (np.typecodes['AllInteger'] + 'S'):
+        # Optimization inspired by <http://stackoverflow.com/a/16973510/325565>
+        dtype = np.dtype((np.void, ar.dtype.itemsize * ar.shape[1]))
+    else:
+        dtype = [('f{i}'.format(i=i), ar.dtype) for i in range(ar.shape[1])]
```

Comment:
Oops, I misread this :).

It would be good to replace the link to the stackoverflow post in the comment with a brief description like the one you add here.

I'm not sure about booleans, but time types all have unique binary representations -- they use int64 under the covers, with the smallest int64 value reserved for NaT.